### PR TITLE
Bazel

### PR DESCRIPTION
**Pull request by fmorency**
MVP

Bazel files are generated from the `Cargo.toml` manifests.

Uses `rules_rust` v0.10.0. Tested using Bazel 5.2.0.

**Supported features**
- Build `many-rs`
- Test `many-rs` (HSM and cucumber included)
- Generate `many-rs` code coverage
- Run fmt
- Run clippy
- Generate rust-analyzer JSON project file

**Supported configurations**
- build:rustfmt         : Enable rustfmt for all targets
- build:clippy          : Enable clippy for all targets
- build:rust-analyzer   : Generate RA project JSON

**TODO**
- DRY (w/ macros)
- Update README
- Adjust rustfmt
- Toggle features from the CLI
- Create a release package

**How-to**
- Sync crates: `CARGO_BAZEL_REPIN=1 bazel sync --only=crate_index`
- Build: `bazel build //...:all-targets`
- Test: `bazel test //...:all-targets`
- Rustfmt: `bazel build --config=rustfmt //...:all-targets`
- Clippy: `bazel build --config=clippy //...:all-targets`
- Rust-analyzer: `bazel build --config=rust-analyzer //...:all-targets`
- Code coverage
```
$ bazel coverage --combined_report=lcov //...:all-targets 
$ genhtml --output genhtml "$(bazel info output_path)/_coverage/_coverage_report.dat"
```

Notes
- I enabled `ecdsa` and `ed25519` by default in `many-identity`
- Each library has a "normal" and a "test" version. Tests always use the "test" version which enables more features

### CODE REVIEW

File: .bazelrc [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+# Enable clippy for all targets in the workspace
+build:clippy --aspects=@rules_rust//rust:defs.bzl%rust_clippy_aspect
+build:clippy --output_groups=+clippy_checks
+
+# Enable rust-analyzer project json for all targets in the workspace
+build:rust-analyzer @rules_rust//tools/rust_analyzer:gen_rust_project
+
+# This import should always be last to allow users to override
+# settings for local development.
+try-import %workspace%/user.bazelrc
```
Comment (hansl) [inline .bazelrc:16]: end of line.

File: src/many-identity-hsm/BUILD.bazel [hunk 2]
```diff
@@ -0,0 +1,50 @@
...
+        normal_dev = True,
+    ) + [
+        "//src/many-error",
+        "//src/many-identity:many-identity-for-test",
+        "//src/many-identity-dsa:many-identity-dsa-for-test",
+        "//src/many-protocol:many-protocol-for-test",
+    ],
+)
+
+rust_test(
```
Comment (hansl) [inline src/many-identity-hsm/BUILD.bazel:45]: Is this working on your machine?

Comment (hansl) [reply src/many-identity-hsm/BUILD.bazel:45]: image

**Review by hansl — CHANGES_REQUESTED** [review]

File: src/many-identity-hsm/BUILD.bazel (hunk 2 above)
Comment (hansl) [reply src/many-identity-hsm/BUILD.bazel:45]: Seems `mock` is flaky, not failing. I got it to fail once above, but now it passes.

For HSM, I get:

```
exec ${PAGER:-/usr/bin/less} "$0" || exit 1
Executing tests from //src/many-identity-hsm:many-identity-hsm-test
-----------------------------------------------------------------------------

running 2 tests
test tests::hsm_ecdsa_sign_verify ... FAILED
test tests::hsm_ecdsa_sign_p256_verify ... FAILED

failures:

---- tests::hsm_ecdsa_sign_verify stdout ----
thread 'tests::hsm_ecdsa_sign_verify' panicked at 'Unable to load PKCS#11 module: LibraryLoading(DlOpen { desc: "dlopen(/usr/lib/softhsm/libsofthsm2.so, 0x0005): tried: \'/usr/lib/softhsm/libsofthsm2.so\' (no such file)" })', src/many-identity-hsm/src/lib.rs:229:50
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace

---- tests::hsm_ecdsa_sign_p256_verify stdout ----
Error: ManyError(Reason { code: HSMMutexPoisoned, message: Some("PKCS#11 global instance mutex poisoned:\n{details}"), arguments: {"details": "poisoned lock: another task failed inside"} })
thread 'tests::hsm_ecdsa_sign_p256_verify' panicked at 'assertion failed: `(left == right)`
  left: `1`,
 right: `0`: the test returned a termination value with a non-zero status code (1) which indicates a failure', /rustc/a8314ef7d0ec7b75c336af2c9857bfaf43002bfc/library/test/src/lib.rs:185:5


failures:
    tests::hsm_ecdsa_sign_p256_verify
    tests::hsm_ecdsa_sign_verify

test result: FAILED. 0 passed; 2 failed; 0 ignored; 0 measured; 0 filtered out; finished in 0.02s
```

Comment (hansl) [reply src/many-identity-hsm/BUILD.bazel:45]: Alright `bazel` needs the `--test_env`. I should have thought about it.

Still no idea why the first mock test failed. Can't repro even when I purge.

**Review by hansl — APPROVED** [review]
LGTM assuming you add a newline at the end of `.bazelrc`.

**Review by Denommus — APPROVED** [review]

**Pull request merged**